### ISSUE TITLE
Unconditional recursion lint doesn't catch operator overloads

### ISSUE THREAD

**Issue by huonw** [rust-lang/rust#23779]
E.g.

```rust
use std::ops::Index;
struct X;
impl Index<usize> for X {
    type Output = ();
    fn index<'a>(&'a self, index: usize) -> &'a () {
        &self[index]
    }
}
fn main() {
    X[0]
}
```

compiles without a peep from the compiler, but will overflow the stack (if optimisations are off). The `self[index]` call is calling `<X as Index<_>>::index`, i.e. exactly the function definition it is inside.

**Comment by huonw** [comment]
This seems to give a warning now, yay!
